cloud/azure: TestAzureFileCredential failed

cloud/azure.TestAzureFileCredential [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20795186?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20795186?buildTab=artifacts#/) on master @ [eff85316e008766c560506b5594dbaf7b2842738](https://github.com/cockroachdb/cockroach/commits/eff85316e008766c560506b5594dbaf7b2842738):


```
=== RUN   TestAzureFileCredential
--- FAIL: TestAzureFileCredential (115.31s)
=== RUN   TestAzureFileCredential/reload-on-error
--- FAIL: TestAzureFileCredential/reload-on-error (0.48s)
=== RUN   TestAzureFileCredential/reload-on-error/invalid-on-reload
    azure_file_credentials_test.go:230: 
        	Error Trace:	pkg/cloud/azure/azure_file_credentials_test.go:230
        	Error:      	An error is expected but got nil.
        	Test:       	TestAzureFileCredential/reload-on-error/invalid-on-reload
--- FAIL: TestAzureFileCredential/reload-on-error/invalid-on-reload (0.02s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAzureFileCredential.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57372

--- a/pkg/cloud/azure/azure_file_credential.go
+++ b/pkg/cloud/azure/azure_file_credential.go
@@ -62,6 +62,10 @@ func NewAzureFileCredential(
 	return c, nil
 }
 
+func credFileEnv() string {
+	return envutil.EnvOrDefaultString("COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE", "")
+}
+
 // NewDefaultAzureCredentialWithFile creates a credential chain that's simply
 // a FileCredential prepended to the chain in DefaultAzureCredential.
 func NewDefaultAzureCredentialWithFile(
@@ -74,7 +78,7 @@ func NewDefaultAzureCredentialWithFile(
 	var credentials []azcore.TokenCredential
 	var credErrors []string
 
-	credentialFileName := envutil.EnvOrDefaultString("COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE", "")
+	credentialFileName := credFileEnv()
 	fileCredentials, err := NewAzureFileCredential(credentialFileName, &FileCredentialOptions{ClientOptions: options.ClientOptions, testingKnobs: options.testingKnobs})
 	if err != nil {
 		credErrors = append(credErrors, errors.Wrap(err, "file credential").Error())

--- a/pkg/cloud/azure/azure_file_credentials_test.go
+++ b/pkg/cloud/azure/azure_file_credentials_test.go
@@ -211,8 +211,7 @@ func TestAzureFileCredential(t *testing.T) {
 
 		t.Run("invalid-on-reload", func(t *testing.T) {
 			credFile := path.Join(tmpDir, "reload-on-error-invalid-on-reload")
-			cleanup := envutil.TestSetEnv(t, "COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE", credFile)
-			defer cleanup()
+			defer envutil.TestSetEnv(t, "COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE", credFile)()
 
 			require.NoError(t, writeAzureCredentialsFile(credFile, cfg.tenantID, cfg.clientID, cfg.clientSecret))
 

--- a/pkg/cloud/azure/azure_storage.go
+++ b/pkg/cloud/azure/azure_storage.go
@@ -28,6 +28,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/envutil"
 	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -222,10 +223,16 @@ type azureStorage struct {
 	settings  *cluster.Settings
 }
 
+type azCacheKey struct {
+	conf    cloudpb.ExternalStorage_Azure
+	envFile string
+	envID   string
+}
+
 var azClientCache struct {
 	syncutil.Mutex
 	// TODO(dt): make this an >1 item cache e.g. add a FIFO ring.
-	key    cloudpb.ExternalStorage_Azure
+	key    azCacheKey
 	set    time.Time
 	client *service.Client
 }
@@ -274,14 +281,19 @@ func makeAzureStorage(
 
 	var azClient *service.Client
 
-	clientConf := *conf
-	clientConf.Prefix = "" // Prefix is not part of the client identity.
+	clientID, _ := envutil.ExternalEnvString("AZURE_CLIENT_ID", 0)
+	cacheKey := azCacheKey{
+		conf:    *conf,
+		envFile: credFileEnv(),
+		envID:   clientID,
+	}
+	cacheKey.conf.Prefix = "" // Prefix is not part of the client identity.
 
 	if reuseSession.Get(&args.Settings.SV) {
 		func() {
 			azClientCache.Lock()
 			defer azClientCache.Unlock()
-			if cached := azClientCache.client; cached != nil && azClientCache.key == clientConf && timeutil.Since(azClientCache.set) < 2*time.Minute {
+			if cached := azClientCache.client; cached != nil && azClientCache.key == cacheKey && timeutil.Since(azClientCache.set) < 2*time.Minute {
 				azClient = cached
 			}
 		}()
@@ -342,7 +354,7 @@ func makeAzureStorage(
 	if reuseSession.Get(&args.Settings.SV) {
 		azClientCache.Lock()
 		defer azClientCache.Unlock()
-		azClientCache.key = clientConf
+		azClientCache.key = cacheKey
 		azClientCache.client = azClient
 		azClientCache.set = timeutil.Now()
 	}

